graph/generate.go

```diff
@@ -13,142 +9,177 @@ import (
+		_, err := g.addMsgLineage(ancestorEnv, ancestor, node, msg, projectors, false)
 		if err != nil {
-			return nil, fmt.Errorf("constructing new node from msg %v failed with error;\n\t%w", ctx.msg, err)
+			return nil, fmt.Errorf("failed to add lineage for message %v:\n%w", ancestor, err)
 		}
+	}
+	return node, nil
+}
 
-		ancestors, err := ancestors(ctx.msg, projectors)
+// addArgLineages adds all the nodes and lineages belonging to argMsgs to the graph as Arguments of a projector.
+// Evaluating the arguments of a projector creates a new environment which is returned at the end.
+// arguments which are ValueSource_ProjectedValues are handled specially.
+func (g Graph) addArgLineages(e *env, argMsgs []proto.Message, projNode Node, projMsg *mbp.ProjectorDefinition, projectors map[string]*mbp.ProjectorDefinition) (*env, error) {
+	argLookup := map[proto.Message]Node{}
+	for _, argMsg := range argMsgs {
+		var argNode Node
+		var err error
+		if projectedValue, ok := getProjectedValue(argMsg); ok { // the argument is a projected value
+			projDef, ok := projectors[projectedValue.GetProjector()] // the corresponding projector definition is the argument
+			if !ok {
+				return nil, fmt.Errorf("could not find projector belonging to projected value %v", projectedValue)
+			}
+			argNode, err = g.addMsgLineage(e, projDef, projNode, projectedValue, projectors, true)
+		} else {
+			argNode, err = g.addMsgLineage(e, argMsg, projNode, projMsg, projectors, true)
+		}
 		if err != nil {
-			return nil, fmt.Errorf("extracting the ancestors of msg %v failed with error;\n\t%w", ctx.msg, err)
+			return nil, fmt.Errorf("failed to add argument %v lineage:\n%w", argMsg, err)
 		}
+		argLookup[argMsg] = argNode
+	}
+	return projectorEnv(e, projMsg, argMsgs, argLookup), nil
+}
 
-		frontier.reversePush(makeMsgContexts(ancestors, node))
-
-		g.Graph[node.ID()] = []IsID{}
-		if ctx.descendant != nil {
-			g.Graph[ctx.descendant.ID()] = append(g.Graph[ctx.descendant.ID()], node.ID())
+func getProjectedValue(msg proto.Message) (*mbp.ValueSource, bool) {
+	if source, ok := msg.(*mbp.ValueSource); ok {
+		if projectedValue, ok := source.GetSource().(*mbp.ValueSource_ProjectedValue); ok {
+			return projectedValue.ProjectedValue, true
 		}
-		g.Nodes[node.ID()] = node
 	}
+	return nil, false
+}
 
-	return &g, nil
+// getNode returns a node given a whistler message.
+// It checks if the node was already created as a projector argument and returns it if so.
+// Otherwise, it creates and returns a new node for the message.
+// It returns as a boolean whether the node was newly created (true) or not (false)
+func getNode(msg proto.Message, descendantNode Node, argLookup map[proto.Message]Node) (Node, bool, error) {
+	if _, ok := descendantNode.(*ArgumentNode); ok {
+		node, ok := argLookup[msg]
+		if !ok {
+			return nil, false, fmt.Errorf("expected msg %v to be in argLookup, but it was not", msg)
+		}
+		return node, false, nil
+	} else {
+		node, err := newNode(msg)
```

Comment:
Do you need to put the created node into the lookup table?

Reply:
The lookup table is only used to find previously generated ancestors of argument nodes. A node whose descendant is not an ArgumentNode doesn't need to be added to the lookup table. The lookup table is removed in the next PR and replaced with whistlerNode, a struct providing extra context to a whistler message, to help make this clearer.